python_modules/dagster/dagster/core/definitions/resource_requirement.py

```diff
@@ -0,0 +1,169 @@
+):
+    @staticmethod
+    def expected_type(self) -> Type:
+        from ..storage.root_input_manager import RootInputManagerDefinition
+
+        return RootInputManagerDefinition
+
+    def describe_requirement(self) -> str:
+        return f"input manager with key '{self.key}' required by input '{self.input_name}' of {self.solid_def.node_type_str} '{self.node_handle.to_string()}'"
+
+
+class OutputManagerRequirement(
+    NamedTuple(
+        "_OutputManagerRequirement", [("key", str), ("node_description", str), ("output_name", str)]
+    ),
+    ResourceRequirement,
+):
+    @property
+    def expected_type(self) -> Type:
+        from ..storage.io_manager import IOManagerDefinition
+
+        return IOManagerDefinition
+
+    def describe_requirement(self) -> str:
+        return f"io manager with key '{self.key}' required by output '{self.output_name}' of {self.node_description}'"
+
+
+class HookResourceRequirement(
+    NamedTuple(
+        "_HookResourceRequirement",
+        [("key", str), ("attached_to", Optional[str]), ("hook_name", str)],
+    ),
+    ResourceRequirement,
+):
+    def describe_requirement(self) -> str:
+        attached_to_desc = f"attached to {self.attached_to}" if self.attached_to else ""
+        return (
+            f"resource with key '{self.key}' required by hook '{self.hook_name}' {attached_to_desc}"
+        )
+
+
+class TypeResourceRequirement(
+    NamedTuple("_TypeResourceRequirement", [("key", str), ("type_display_name", str)]),
+    ResourceRequirement,
+):
+    def describe_requirement(self) -> str:
+        return f"resource with key '{self.key}' required by type '{self.type_display_name}'"
+
+
+class TypeLoaderResourceRequirement(
+    NamedTuple("_TypeLoaderResourceRequirement", [("key", str), ("type_display_name", str)]),
+    ResourceRequirement,
+):
+    def describe_requirement(self) -> str:
+        return f"resource with key '{self.key}' required by the loader on type '{self.type_display_name}'"
+
+
+class TypeMaterializerResourceRequirement(
+    NamedTuple("_TypeMaterializerResourceRequirement", [("key", str), ("type_display_name", str)]),
+    ResourceRequirement,
+):
+    def describe_requirement(self) -> str:
+        return f"resource with key '{self.key}' required by the materializer on type '{self.type_display_name}'"
+
+
+class ResourceDependencyRequirement(
+    NamedTuple("_ResourceDependencyRequirement", [("key", str), ("source_key", str)]),
+    ResourceRequirement,
+):
+    def describe_requirement(self) -> str:
+        return f"resource with key '{self.key}' required by resource with key '{self.source_key}'"
+
+
+class RequiresResources(ABC):
+    @property
+    @abstractmethod
+    def required_resource_keys(self) -> AbstractSet[str]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_resource_requirements(
+        self, outer_context: Optional[object] = None
+    ) -> Iterator[ResourceRequirement]:
+        raise NotImplementedError()
```

Comment:
This piece is definitely weird. We only need it for error messages, and in all the implementations here it's just a string. But theoretically can support more richly structured information.